File: 125-valid-palindrome/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        p1 , p2  = 0 , len(s)-1
        while p1 < p2:
            if not s[p1].isalnum():
                p1 += 1
                continue
            if not s[p2].isalnum():
                p2 -= 1
                continue
            if s[p1] == s[p2]:
                p1 += 1
                p2 -= 1
                continue
            else:
                return False
        return True

File: 125-valid-palindrome/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            char lChar = s.charAt(left);
            char rChar = s.charAt(right);

            // skip non-alphanumeric characters
            if (!Character.isLetterOrDigit(lChar)) {
                left++;
                continue;
            }
            if (!Character.isLetterOrDigit(rChar)) {
                right--;
                continue;
            }

            // compare ignoring case
            if (Character.toLowerCase(lChar) != Character.toLowerCase(rChar)) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }
}
